compiler/hash-lower/src/ty/new.rs

```diff
@@ -0,0 +1,263 @@
+                                let fields = fields
+                                    .iter()
+                                    .map(|field| {
+                                        let ty = self.ty_id_from_tir_ty(field.ty);
+                                        let name = self
+                                            .get_symbol(field.name)
+                                            .name
+                                            .unwrap_or(IDENTS.underscore);
+                                        // @@Verify: do we need to store default parameter
+                                        // information here?
+                                        AdtField { name, ty }
+                                    })
+                                    .collect::<Vec<_>>();
+
+                                AdtVariant { name, fields }
+                            })
+                        })
+                        .collect::<AdtVariants>()
+                });
+
+                // Get the name of the data type, if no name exists we default to
+                // using `_`.
+                let name = self.get_symbol(data_ty.name).name.unwrap_or(IDENTS.underscore);
+                let adt = AdtData::new_with_flags(name, variants, flags);
+
+                let id = self.lcx.adts().create(adt);
+                IrTy::Adt(id)
+            }
+
+            // Primitive are numerics, strings, arrays, etc.
+            DataDefCtors::Primitive(primitive) => {
+                match primitive {
+                    PrimitiveCtorInfo::Numeric(NumericCtorInfo { bits, is_signed, is_float }) => {
+                        if is_float {
+                            match bits {
+                                NumericCtorBits::Bounded(32) => IrTy::Float(FloatTy::F32),
+                                NumericCtorBits::Bounded(64) => IrTy::Float(FloatTy::F64),
+
+                                // Other bits widths are not supported.
+                                _ => unreachable!(),
+                            }
+                        } else {
+                            match bits {
+                                NumericCtorBits::Bounded(bits) => {
+                                    let size = Size::from_bits(bits);
+
+                                    if is_signed {
+                                        IrTy::Int(SIntTy::from_size(size))
+                                    } else {
+                                        IrTy::UInt(UIntTy::from_size(size))
+                                    }
+                                }
+                                // We don't implement bigints yet
+                                _ => unimplemented!(),
+                            }
+                        }
+                    }
+
+                    // @@Verify: does `str` now imply that it is a `&str`, or should we create a
+                    // `&str` type here.
```

Comment:
Probably the latter, maybe better to do this once TC deals with it?

Reply:
As in that it verifies that the `str` does not appear in a type position (except generic parameters)?

Reply:
Yeah, surely that should be part of the layout logic?